keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,903 @@
+may want to control creation of volumes carefully in order to optimize
+availability, durability, performance and cost. For more information
+about this, see the [StoragePool API for Advanced Storage Placement
+KEP](https://github.com/kubernetes/enhancements/pull/1347).
+
+Storage pools as introduced in this KEP enable the creation of such
+operators by providing them the necessary information about storage in
+the cluster.
+
+### Caching remaining capacity via the API server
+
+CSI defines the `GetCapacity` RPC call for the controller service to
+report information about the current capacity but it's not accessible
+in the Kubernetes scheduler. A new type `CSIStoragePool` type gets
+added to store the result of the `GetCapacity` calls for one storage
+pool in the API server and thus make it available to the scheduler. In
+the future it might get extended to also store other information like
+health of the pool.
+
+### Identifying storage pools
+
+A method for identifying storage pools accessible via a CSI driver
+based on the existing topology information gets outlined below. The
+`external-provisioner` will be extended to handle the management of
+`CSIStoragePool` objects.
+
+For local storage in the node, there will be one such `CSIStoragePool`
+object per node and CSI driver on that node. For network attached
+storage, there will be one `CSIStoragePool` per CSI
+[Topology](https://github.com/container-storage-interface/spec/blob/4731db0e0bc53238b93850f43ab05d9355df0fd9/lib/go/csi/csi.pb.go#L1662-L1691)
+and driver.
+
+### Size of ephemeral inline volumes
+
+Currently Kubernetes has no information about the size of an ephemeral
+inline volume. A new `CSIVolumeSource.fsSize` field needs to be added
+to expose that in a vendor-agnostic way. Details for that are in
+https://github.com/kubernetes/enhancements/pull/1353.
+
+### Pod scheduling
+
+The Kubernetes scheduler monitors `CSIStoragePool` objects and
+excludes nodes with insufficient remaining capacity when it comes to
+making scheduling decisions for a pod which uses ephemeral inline
+volumes or persistent volumes with delayed binding. If the driver does
+not indicate that it supports capacity reporting, then the scheduler
+proceeds just as it does now, so nothing changes for existing CSI
+driver deployments.
+
+### API
+
+All Kubernetes API changes are under the `CSIStorageCapacity` feature
+gate.
+
+#### CSIStoragePool
+
+The prefix “CSI” is used to avoid confusion with other types that have
+a similar name and because some aspects are specific to CSI.
+
+```
+// CSIStoragePool identifies one particular storage pool and
+// stores the corresponding attributes. The spec is read-only.
+//
+// There should never be two objects with the same spec. If
+// a consumer of this information finds two objects with the
+// same spec, it should use the information from the one with the
+// most recent CreationTimestamp.
+type CSIStoragePool struct {
```

Comment:
Specify that CSIStoragePool is non-namespaced.

Reply:
Ack.